File: study/lua/src/app/ui/component/CommonTacticsIcon.lua
--
-- Author: Wangyu
-- Date: 
-- 战法Icon

local CommonIconBase = import(".CommonIconBase")
local CommonTacticsIcon = class("CommonTacticsIcon", CommonIconBase)
local ComponentIconHelper = require("app.ui.component.ComponentIconHelper")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local EffectGfxNode = require("app.effect.EffectGfxNode")
local PopupTacticsDetail = require("app.ui.PopupTacticsDetail")

local EXPORTED_METHODS = {
    "updateUI"
}

function CommonTacticsIcon:ctor()
    CommonTacticsIcon.super.ctor(self)
    self._type = TypeConvertHelper.TYPE_TACTICS
end

function CommonTacticsIcon:_init()
    CommonTacticsIcon.super._init(self)
end


function CommonTacticsIcon:bind(target)
    CommonTacticsIcon.super.bind(self, target)

    cc.setmethods(target, self, EXPORTED_METHODS)
end

function CommonTacticsIcon:unbind(target)
    CommonTacticsIcon.super.unbind(self, target)

    cc.unsetmethods(target, EXPORTED_METHODS)
end

--根据传入参数，创建并，更新UI
function CommonTacticsIcon:updateUI(value, size, rank, limitLevel, limitRedLevel)
    local itemParams = CommonTacticsIcon.super.updateUI(self, value, size, rank, limitLevel, limitRedLevel)
    return itemParams
end

function CommonTacticsIcon:_onTouchCallBack(sender, state)
    -----------防止拖动的时候触发点击
    if (state == ccui.TouchEventType.ended) then
        local moveOffsetX = math.abs(sender:getTouchEndPosition().x - sender:getTouchBeganPosition().x)
        local moveOffsetY = math.abs(sender:getTouchEndPosition().y - sender:getTouchBeganPosition().y)
        if moveOffsetX < 20 and moveOffsetY < 20 then
            if self._callback then
                self._callback(sender, self._itemParams)
            else
                local baseId = self._itemParams.cfg.id
                local popup = PopupTacticsDetail.new(sender, baseId)
                popup:open()
            end
        end
    end
end

return CommonTacticsIcon

File: study/ts/script/ui/component/CommonTacticsIcon.ts
import CommonIconBase from "./CommonIconBase";
import { TypeConvertHelper } from "../../utils/TypeConvertHelper";
import { G_SceneManager } from "../../init";
import PopupTacticsDetail from "../popup/PopupTacticsDetail";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CommonTacticsIcon extends CommonIconBase {
    /**
     *
     */
    constructor() {
        super();
        this._type = TypeConvertHelper.TYPE_TACTICS;
    }
    _init() {
        CommonTacticsIcon.prototype._init.call(this);
    }
    updateUI(value, size?, rank?, limitLevel?, limitRedLevel?) {
        var itemParams = super.updateUI(value, size, rank, limitLevel, limitRedLevel);
        return itemParams;
    }
    _onTouchCallBack(sender, state) {
        if (state == cc.Node.EventType.TOUCH_END) {
            var moveOffsetX = Math.abs(sender.getTouchEndPosition().x - sender.getTouchBeganPosition().x);
            var moveOffsetY = Math.abs(sender.getTouchEndPosition().y - sender.getTouchBeganPosition().y);
            if (moveOffsetX < 20 && moveOffsetY < 20) {
                if (this._callback) {
                    this._callback(sender, this._itemParams);
                } else {
                    var baseId = this._itemParams.cfg.id;
                    G_SceneManager.openPopup('prefab/common/PopupTacticsDetail', (popup: PopupTacticsDetail) => {
                        popup.ctor(sender, baseId);
                        popup.open();
                    })
                }
            }
        }
    }
}